### PR TITLE
Enable dev services for all named mongodb-client connections

### LINKED ISSUE #25278: Dev Services for quarkus-mongo-client don't start if @MongoClientName used
### Describe the bug

Quarkus doesn't try to start MongoDB in dev mode if you inject `MongoClient` with `@MongoClientName`.
There's nothing in logs (with debug logging enabled) about reason why Quarkus did't start db.

### Expected behavior

Dev service started and same `MongoClient` injected for all injection points.

### Actual behavior

Quarkus silently skips starting mongodb container and ArC fails to create bean.

### How to Reproduce?

```
quarkus create app -x resteasy-reactive -x mongodb-client
# edit src/main/java/org/acme/GreetingResource.java to include @Inject @MongoClientName("test") MongoClient client; 
# and invoke something on the client like client.getDatabase("test")
quarkus dev
# curl http://localhost:8080/hello
```

You'll get something like this:
```
java.lang.RuntimeException: mongo config is missing for creating mongo client.
        at io.quarkus.mongodb.runtime.MongoClients.createMongoConfiguration(MongoClients.java:253)
        at io.quarkus.mongodb.runtime.MongoClients.createMongoClient(MongoClients.java:118)
```

### Output of `uname -a` or `ver`

Linux unterwelt 5.17.4-arch1-1 #1 SMP PREEMPT Wed, 20 Apr 2022 18:29:28 +0000 x86_64 GNU/Linux

### Output of `java -version`

OpenJDK 64-Bit Server VM Temurin-17.0.2+8 (build 17.0.2+8, mixed mode, sharing)

### GraalVM version (if different from Java)

N/A

### Quarkus version or git rev

2.8.2.Final

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Gradle 7.4.1

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by evanchooly**
fixes #25278

### CODE REVIEW

**Comment by geoand** [comment]
> Could we infer the existence of a named client by discovering uses of @MongoClientName and backfilling that client using only that name and the default values of the config? In essence, lazy init'ing the named client and letting devservices have its usual go at it

Probably, but I am thinking that it might break stuff, so I would like to avoid it.

File: integration-tests/mongodb-client/src/main/java/io/quarkus/it/mongodb/BookResource.java [hunk 1]
```diff
@@ -9,17 +9,20 @@
...
 
+import io.quarkus.mongodb.MongoClientName;
 import io.smallrye.common.annotation.Blocking;
 
 @Path("/books")
 @Blocking
 public class BookResource {
 
     @Inject
+    @MongoClientName("parameter-injection")
```
Comment (geoand) [inline integration-tests/mongodb-client/src/main/java/io/quarkus/it/mongodb/BookResource.java:25]: I think we need the test to have both named and non-named clients

**Comment by evanchooly** [comment]
> > Could we infer the existence of a named client by discovering uses of @MongoClientName and backfilling that client using only that name and the default values of the config? In essence, lazy init'ing the named client and letting devservices have its usual go at it
> 
> Probably, but I am thinking that it might break stuff, so I would like to avoid it.

actually, the more i think about this the less i like it.  typos would be a nightmare.  (even though those names should be held in constants somewhere. ;) )

**Comment by evanchooly** [comment]
*bump*

**Comment by geoand** [comment]
The affected test is **not** using dev services as the connection string has been set - mongo is being spun up by `io.quarkus.test.mongodb.MongoTestResource`

File: integration-tests/mongodb-client/src/main/resources/application.properties [hunk 2]
```diff
@@ -1,4 +1,7 @@
 quarkus.mongodb.connection-string=mongodb://localhost:27017
 quarkus.mongodb.write-concern.journal=false
 quarkus.mongodb.parameter-injection.connection-string=mongodb://localhost:27017
-quarkus.mongodb.parameter-injection.write-concern.journal=false
\ No newline at end of file
+quarkus.mongodb.parameter-injection.write-concern.journal=true
+
+quarkus.mongodb.dev-services.write-concern.journal=false
+quarkus.mongodb.dev-services2.write-concern.journal=false
```
Comment (evanchooly) [inline integration-tests/mongodb-client/src/main/resources/application.properties:7]: what do you think of this @geoand?  And the `BookResource` change above.  It's not the "cleanest" but I was trying to find a simple way to test both the `MongoTestResource` driven stuff and the dev services as well and this is the simplest I could make it.

Comment (geoand) [reply integration-tests/mongodb-client/src/main/resources/application.properties:7]: I like the idea.

File: integration-tests/mongodb-client/src/main/java/io/quarkus/it/mongodb/BookResource.java [hunk 3]
```diff
@@ -9,36 +9,60 @@
...
+                books.add(doc);
+            }
         }
         return books;
     }
 
     @POST
     public Response addBook(Book book) {
         getCollection().insertOne(book);
+        getServicedCollection().insertOne(book);
```
Comment (geoand) [inline integration-tests/mongodb-client/src/main/java/io/quarkus/it/mongodb/BookResource.java:64]: We aren't querying this collection anywhere, are we? I think we should close the loop (somewhere during the duration of the test) and also do that

Comment (evanchooly) [reply integration-tests/mongodb-client/src/main/java/io/quarkus/it/mongodb/BookResource.java:64]: smart.  just pushed that change.

**Review by geoand — APPROVED** [review]
LGTM!

**Comment by evanchooly** [comment]
> Hm... The test failure is suspicious

No docker on CI ...

**Comment by loicmathieu** [comment]
@evanchooly there is no docker on Windows CI. That's why we don't use testcontainers for MongoDB tests but the testresource. But even the testresource didn't works very well on Windows due to some issue on flapdoodle side so we ends up disabling the test on Windows: `@DisabledOnOs(WINDOWS)`.

**Comment by loicmathieu** [comment]
> @evanchooly there is no docker on Windows CI. That's why we don't use testcontainers for MongoDB tests but the testresource. But even the testresource didn't works very well on Windows due to some issue on flapdoodle side so we ends up disabling the test on Windows: @DisabledOnOs(WINDOWS).

But you didn't add new tests so I don't understand why you have this ...

**Comment by geoand** [comment]
> But you didn't add new tests so I don't understand why you have this ...

Probably because:

```xml
            <plugin>
                <artifactId>maven-failsafe-plugin</artifactId>
                <configuration>
                    <skip>true</skip>
                </configuration>
            </plugin>
```

was removed

**Comment by evanchooly** [comment]
fixed the windows failure by disabling it via a profile when on windows.  the last CI run was all green so I squashed the commits and force pushed.  we should be all good to go now.

**Comment by famod** [comment]
It seems this PR has broken `NamedMongoClientConfigTest` and `NamedReactiveMongoClientConfigTest`.

The failsafe-skip-profile won't work for those as both are run via surefire.

**Comment by geoand** [comment]
Yup, @evanchooly is looking into it

**Pull request merged**